Theme-address v2 using country-service

nimzco opened the pull request:
## Problem

In every country, addresses and address forms are rendered differently. 
- Field labels can vary from a country to another (Eg.: `Province` vs `State`)
- Order of the fields will vary from a country to another.
- Country and province names need to be translated

See gif bellow for an example.

Our [country-service](https://github.com/Shopify/country-service) service gives us all the information that we need to fix these problems.

## Solution

Rewrote `theme-address` to take care of all the problem listed above
- Fetch data from country-service given a locale.
- Reorder fields based on the selected country.
- Update label names based on the selected country
- Add event listener to update order and label names when country changes.
- Provide CSS example to take care of reordering (see README)

## Demo

![example](https://user-images.githubusercontent.com/445045/52973395-f9725080-338b-11e9-8424-9418a551be18.gif)

nimzco commented on packages/theme-addresses/loader.ts:
```diff
@@ -0,0 +1,97 @@
+}
+
+var DEFAULT_LOCALE = 'EN';
+export var SUPPORTED_LOCALES = [
```
That part will go away pretty soon. We'll handle that on the backend. cc @froyomuffin

froyomuffin replied:
T-T

froyomuffin replied:
Handled

wizardlyhel commented:
Conversation from before - This library would need a way to accept a limited list of countries.

**Scenario:** A template is displaying both billing address and shipping address. 
A billing address can be anywhere in the world but a shipping address is limited to where the merchant can ship their products to.

Since the country-service is a static api, it is better to request once and reuse for multiple times for generating the required address form according to need.

This would require a new liquid tag that outputs just the shipping address country codes.

nimzco commented:
> **Scenario**: A template is displaying both billling address and shipping address.
> A billing address can be anywhere in the world but a shipping address is limited to where the merchant can ship their products to.

Currently, there is no restriction in the theme between a shipping and billing address when you add an address. Since the scenario you're describing is not handled right now we probably want to keep all of the countries in the dropdown as we do here, shouldn't we?

t-kelly commented on packages/theme-addresses/addressForm.ts:
```diff
@@ -0,0 +1,274 @@
+ * Sets up event listener for country change
+ */
+function setEventListeners(rootEl: HTMLElement, formElements: FormElements, country: Country) {
+  formElements.country.input.addEventListener('change', function(event) {
```
You addEventListener whenever the country selector changes. Doesn't this result in multiple event handlers being built up and repeating the task more than needed?

nimzco replied:
Oh true, good catch 👍

t-kelly commented on packages/theme-addresses/addressForm.ts:
```diff
@@ -0,0 +1,274 @@
+var COUNTRIES: Country[];
+var FIELD_REGEXP = /({\w+})/g;
+var LINE_DELIMITER = '_';
+var INPUT_SELECTORS = {
```
These could be overrided via options?

nimzco replied:
Yep, why not. 👍

t-kelly commented on packages/theme-addresses/addressForm.ts:
```diff
@@ -0,0 +1,274 @@
+    var optionElement = document.createElement('option');
+    optionElement.value = country.code;
+    optionElement.textContent = country.name;
+    countrySelect.appendChild(optionElement);
```
You should create a duplicate of `countrySelect` in memory, append the new options to it, and then replace the version of `countrySelect` in the DOM with your updated one. This will significantly reduce the number of DOM operations you are performing.

t-kelly commented on packages/theme-addresses/addressForm.ts:
```diff
@@ -0,0 +1,274 @@
+      }
+
+      if (afterCountry) {
+        rootEl.append(formElements[line].wrapper);
```
You should create a duplicate of rootEl in memory, append the newly ordered fields to it, and then replace the version of rootEl in the DOM with your updated one. This will significantly reduce the number of DOM operations you are performing.

nimzco replied:
👍

nimzco replied:
Actually in this case it fails on the `insertBefore` bellow because I'm dealing with in memory and in the DOM element I think.
But I only have ~10 operations here so we should be good in terms of performance. I'll change it for the country `select` though.

t-kelly commented on packages/theme-addresses/addressForm.ts:
```diff
@@ -0,0 +1,274 @@
+    var optionElement = document.createElement('option');
+    optionElement.value = zone.code;
+    optionElement.textContent = zone.name;
+    zoneSelect.appendChild(optionElement);
```
Same here with DOM operations.

nimzco commented on packages/theme-addresses/__tests__/theme-addresses.test.js:
```diff
@@ -1,232 +1,121 @@
+      countries
+    );
+    AddressForm(document.body.querySelector('[data-address="root"]'), 'en', {
+      loaded,
```
It's for the tests to start running only when the `AddressForm` has finished loading. 
When you pass the `loaded` argument in the `beforeAll` it'll wait for it to be executed before running the tests. It's good for async functions.

nimzco commented on packages/theme-addresses/addressForm.ts:
```diff
@@ -0,0 +1,274 @@
+  if (COUNTRIES) {
+    init(rootEl, formElements, options);
+  } else {
+    loadCountries(locale).then(function(countries) {
```
Oh true... 👍

wizardlyhel commented:
> Currently, there is no restriction in the theme between a shipping and billing address when you add an address. Since the scenario you are describing is not handled right now we probably want to keep all of the countries in the dropdown as we do here, shouldn't we?

I honestly don't know how shipping country liquid tag is being used by the partners either but not supporting it would be like no backward compatibility for shipping address form in particular. Since this PR completely changes the way a theme developer renders an address form, we should be responsible of offering the equivalent to what we already offer from liquid renders. (@nimzco I'll slack you the query report of the tag usage across Shopify)

antoinegrant commented:
> I'm hesitant about the introduction of TypeScript in the repo. I don't have any previous experience with it (and I don't know the experience level of other maintainers). So I'm not sure what we're signing up for by merging it 🙃
> 
> Given I don't know what the learning curve is like, I'm concerned whether it's a realistic expectation of maintainers to support it going forward. I would appreciate it if we could hold off on merging until we've checked in with some of the other repo maintainers about the decision. (We can do this over Slack) Maybe they're all stoked and I have no need to be worried 😄
> 
> All in all : This is API is a great feature @nimzco. I just want to make sure we're ready to maintain it.

I agree with Nathan's point. I would also favour consistency in patterns throughout theme-scripts for simpler global configurations. TypeScript is a great idea for theme-scripts, I'd convert all packages at once as it's own project.

froyomuffin commented:
We'll have the same issue as we have today with CanadaPod restrictions right?

nimzco commented:
Yes... 😭

nimzco commented:
Thanks for your feedback folks. I've removed all the TypeScript from the code and left everything as JS.
But as @froyomuffin mentioned, merchants on the Canada pod shouldn't use this feature...

Also, I haven't taken in account @wizardlyhel's comment on passing it a list of restricted country codes. I've got to dig into that since we don't have access easily to a list like that.

NathanPJF commented:
Given @nimzco's comment https://github.com/Shopify/theme-scripts/pull/78#issuecomment-475617508, it looks like there are known (temporary?) Shopify issues that would prevent this from working for all merchants. 

In broad strokes, a Theme Store theme shouldn't have functionality that only works for a subset of merchants because those themes are intended to work everywhere. However, if there is a graceful fallback in place that's easy for developers to know when to use, that's a different story.

We try to have graceful fallbacks in web development all the time. i.e. No JS flows, reasonable layout if there's no support for CSS Grid.  A fallback in the context of this PR could be to detect whether the feature is available, and if not, fallback to the hardcoded Liquid way that exists now.  @nimzco any reason to suspect that this couldn't happen with using this feature?

Now, it must be said that this fallback approach feels a little weird given it's not going to be clear to merchants and agencies why certain stores are getting it.  Meaning: Outside developers and merchants don't know what "CanadaPod" is and that their expectations should change because the store is hosted there.  This could be partly addressed with public developer documentation about the feature you created @nimzco with a disclaimer that this isn't available across the board.

nimzco commented:
Hey folks!

I've updated the code to make it work for shipping countries only. I've added a boolean option that, when passed, I query the shop's` /meta.json` endpoint where I can retrieve the list of countries the shop ships to. Eg.: https://www.chubbiesshorts.com/meta.json

As for the fallback on no-js, currently, on Minimal & Debut (the only one I've tested), you can't add or edit an address (the form won't show since it's toggled with JS).
I would suggest we move forward with this theme script without fallback for now it's that's ok with you.

As of CanadaPod, there is only one merchant concerned about it. I'm in contact with the CanadaPOD team and I will reach out to their MSM to make sure they don't use that script.


Other than that, it's good to 👀 🚢 on my end! 🚀

wizardlyhel commented on packages/theme-addresses/helpers.js:
```diff
@@ -0,0 +1,16 @@
+      }
+    }
+  }
+  return to;
```
I think you can use `Object.assign({}, obj1, obj2)` here https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/assign

Unless, we need deep copy?

nimzco replied:
I did that for browser compatibility, yeah. :/

nimzco commented:
@NathanPJF @wizardlyhel FYI for CanadaPOD: https://github.com/Shopify/trust-advisors/issues/48
